File: test/benchmark/kostya/brainfuck.js
// Kostya Benchmark: brainfuck (Node.js)
// Brainfuck interpreter — interprets a BF program 10000 times
'use strict';

function buildJumpTable(prog) {
    const len = prog.length;
    const jumps = new Int32Array(len);
    const stack = [];
    for (let i = 0; i < len; i++) {
        const c = prog.charCodeAt(i);
        if (c === 91) { // '['
            stack.push(i);
        } else if (c === 93) { // ']'
            const j = stack.pop();
            jumps[i] = j;
            jumps[j] = i;
        }
    }
    return jumps;
}

function runBf(prog, jumps) {
    const len = prog.length;
    const tape = new Uint8Array(30000);
    let dp = 0;
    let ip = 0;
    const output = [];

    while (ip < len) {
        const op = prog.charCodeAt(ip);
        if (op === 43) { // '+'
            tape[dp] = (tape[dp] + 1) & 0xFF;
        } else if (op === 45) { // '-'
            tape[dp] = (tape[dp] + 255) & 0xFF;
        } else if (op === 62) { // '>'
            dp++;
        } else if (op === 60) { // '<'
            dp--;
        } else if (op === 46) { // '.'
            output.push(String.fromCharCode(tape[dp]));
        } else if (op === 91) { // '['
            if (tape[dp] === 0) {
                ip = jumps[ip];
            }
        } else if (op === 93) { // ']'
            if (tape[dp] !== 0) {
                ip = jumps[ip];
            }
        }
        ip++;
    }
    return output.join('');
}

function main() {
    const prog = "++++++++[>++++[>++>+++>+++>+<<<<-]>+>+>->>+[<]<-]>>.>---.+++++++..+++.>>.<-.<.+++.------.--------.>>+.>++.";
    const jumps = buildJumpTable(prog);

    let output = "";
    for (let iter = 0; iter < 10000; iter++) {
        output = runBf(prog, jumps);
    }
    process.stdout.write(output + "\n");
}

main();

File: test/benchmark/kostya/python/brainfuck.py
#!/usr/bin/env python3
# Kostya Benchmark: brainfuck (Python)
# Brainfuck interpreter — interprets a BF program 10000 times
import time


def build_jump_table(prog):
    jumps = [0] * len(prog)
    stack = []
    for i, c in enumerate(prog):
        if c == '[':
            stack.append(i)
        elif c == ']':
            j = stack.pop()
            jumps[i] = j
            jumps[j] = i
    return jumps


def run_bf(prog, jumps):
    tape = bytearray(30000)
    dp = 0
    ip = 0
    output = []
    prog_len = len(prog)
    while ip < prog_len:
        op = prog[ip]
        if op == '+':
            tape[dp] = (tape[dp] + 1) & 0xFF
        elif op == '-':
            tape[dp] = (tape[dp] - 1) & 0xFF
        elif op == '>':
            dp += 1
        elif op == '<':
            dp -= 1
        elif op == '.':
            output.append(chr(tape[dp]))
        elif op == '[':
            if tape[dp] == 0:
                ip = jumps[ip]
        elif op == ']':
            if tape[dp] != 0:
                ip = jumps[ip]
        ip += 1
    return ''.join(output)


def main():
    prog = "++++++++[>++++[>++>+++>+++>+<<<<-]>+>+>->>+[<]<-]>>.>---.+++++++..+++.>>.<-.<.+++.------.--------.>>+.>++."
    jumps = build_jump_table(prog)

    t0 = time.perf_counter_ns()
    output = ""
    for _ in range(10000):
        output = run_bf(prog, jumps)
    t1 = time.perf_counter_ns()

    print(output)
    print(f"__TIMING__:{(t1 - t0) / 1e6:.3f}")


main()
